[SYCL] Enable async_work_group_copy for scalar and vector bool types

v-klochkov opened the pull request:
Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

s-kanaev commented on sycl/include/CL/sycl/group.hpp:
```diff
@@ -274,58 +274,99 @@ template <int Dimensions = 1> class group {
-                                     size_t numElements) const {
+  detail::enable_if_t<!detail::is_bool<dataT>::value, device_event>
+  async_work_group_copy(global_ptr<dataT> dest, local_ptr<dataT> src,
+                        size_t numElements, size_t destStride) const {
```
Is it by-specification that this function uses destination stride while the previous one has source stride?

v-klochkov replied:
Yes, the stride is specified for the global memory argument. 
If Dest is global mem, then Stride is for Dest. Otherwise, it is for Src.

s-kanaev commented on sycl/test/basic_tests/group_async_copy.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+int main() {
+  for (int Stride = 1; Stride < WorkGroupSize; Stride++) {
```
Is there any rationale behind not testing the copy device for strides longer than WG size?

v-klochkov replied:
Here it is not bigger than WG size only to not complicate the initialization and correctness checks.

s-kanaev commented on sycl/test/basic_tests/group_async_copy.cpp:
```diff
@@ -0,0 +1,150 @@
+  auto Acc = Buf.template get_access<access::mode::write>();
+  for (size_t I = 0; I < Buf.get_count(); I += WorkGroupSize) {
+    for (size_t J = 0; J < WorkGroupSize; J++)
+      Acc[I + J] = I + J + ((J % Stride == 0) ? 100 : 0);
```
NIT: wow, it's interesting that `Stride` is nil by default. Yet it is a divider here.

v-klochkov replied:
The initialization was initially written differently, I forgot to remove that default initialization with 0. Will be fixed.

s-kanaev commented on sycl/test/basic_tests/group_async_copy.cpp:
```diff
@@ -0,0 +1,150 @@
+
+template <typename T> int checkResults(buffer<T, 1> &OutBuf, size_t Stride) {
+  auto Out = OutBuf.template get_access<access::mode::read>();
+  int EarlyFailout = 20;
```
How did we get 20 here? Does early failout helps in any way here i.e. in testing?
Is 20 an arbitrary number less than `NElems`?

v-klochkov replied:
20 is an arbitrary number, it is reasonable big enough to let the error log observer see the pattern and have at least some initial guess on what went wrong with the copy even before running the test locally.

s-kanaev commented on sycl/test/basic_tests/group_async_copy.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+int main() {
+  for (int Stride = 1; Stride < WorkGroupSize; Stride++) {
```
Are there any possible negative scenarios?

v-klochkov replied:
I'd rather keep this test as is now without additional changes in initialization/correctness-checks, if you don't mind.

v-klochkov replied:
BTW, by "negative scenarios", did you mean negative Stride?
Currently, SYCL SPEC specifies the Stride arg as unsigned, i.e. size_t
59 template <typename dataT>
60 device_event async_work_group_copy(local_ptr<dataT> dest,
61 global_ptr<dataT> src, size_t numElements, size_t srcStride) const;
62

s-kanaev replied:
> BTW, by "negative scenarios", did you mean negative Stride?

Not really. "Negative" scenario is the one that should result negative or with failure i.e. when copy operation isn't allowed or shouldn't succeed.

s-kanaev reviewed (dismissed):
LGTM overall.
Only got a some questions.

v-klochkov commented:
@s-kanaev , Thank you for the quick code-review.
I made two additional trivial fixes in LIT test now.
- removed default initialization of 'Stride' arg in the initialization function.
- added the checks for async_work_group_copy versions that do not have 'Stride' arg.
